text/0000-pack-buildpack-new-alternatives.md

```diff
@@ -0,0 +1,206 @@
+
+Replacing `pack buildpack new` with `pack buildpack create` allows buildpack authors to more-easily create new buildpacks in their chosen language and framework.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Project scaffolding is [very](https://cookiecutter.readthedocs.io/) [popular](https://yeoman.io/) [in](https://github.com/kowainik/summoner) [other](https://crates.io/crates/cargo-scaffold) [ecosystems](https://github.com/golang-standards/project-layout).  Scaffolding systems are employed to ease onboarding of new developers.  Within `pack` this feature is targeted at onboarding buildpack authors.
+
+`pack buildpack create` should prompt the end user to choose an implementation language.  Thereafter, generate skeleton code according to the best practices for that language.  For example, users choosing Go as an implementation language would generate a [golang standard project layout](https://github.com/golang-standards/project-layout) and include `libcnb` best practices.  The layout would look similar to the following:
+
+```bash
+.
+├── buildpack.toml
+├── cmd
+│   └── main.go
+├── go.mod
+├── go.sum
+└── pkg
+    ├── build.go
+    └── detect.go
+```
+
+This approach is not opinionated on topics such as which testing framework to use.  Only components under github.com/buildpacks/* are used in the generated project.  `pack buildpack create --template URL` allows for more opinionated project scaffolding.  As such, we support new buildpack authors with onboarding and support experienced buildpack authors to use the scaffolding of their choice.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The design is modeled on [cookiecutter](https://cookiecutter.readthedocs.io/en/1.7.2/) with reference to [springerle](https://github.com/carlmjohnson/springerle) -- a similar implementation in golang -- and [create-go-app](https://github.com/create-go-app/).  We do we want to `os.Exec` a python subprocess to run cookiecutter as this would require availability of a python runtime environment.  Instead we propose to borrow heavily from create-go-app, and generate the default shell and libcnb skeletons from a cloned git repoistory.
+
+A full invocation to generate `bash` scaffolding prompts for the values [currently documented as flags](https://buildpacks.io/docs/buildpack-author-guide/create-buildpack/building-blocks-cnb/#using-the-pack-cli).
+
+```bash
+pack buildpack new
+Use the arrow keys to navigate: ↓ ↑ → ←
+? Choose a project template:
+  ▸ bash
+    go
+
+```
+
+A full session includes the terminal prompts that the project scaffolding tool asks of the end user:
+
+```bash
+pack buildpack new
+Use the arrow keys to navigate: ↓ ↑ → ←
+? Choose a project template:
+  ▸ bash
+    go
+✔ Enter a directory in which to scaffold the project: bash_buildpack
+Use the arrow keys to navigate: ↓ ↑ → ←
+? Choose the buildpack API version (use the default if you are unsure):
```

Comment:
This might be better as a free form entry, but default to the latest current release, or have an "other" option. Otherwise, you'd have to update to be able to select newer versions. It would also me we have to bump this every time there's a new version.

Reply:
One possible implementation [scafall](https://github.com/AidanDelaney/scafall/blob/main/pkg/scafall.go#L18) allows the calling code to offer `DefaultValues`.  Using this, `pack` version `0.27` could call `scafall` passing in `BuildpackApi =[0.5, 0.6, 0.7, 0.8]` and some future version of pack could call `scafall` passing in `BuildpackApi =[1.0]`.  This give the calling code some control over the prompts presented to the end-user.

Reply:
So I have pack installed through a package manager and it auto updates, but not everyone does that. Some people are running older versions of pack. It would not be the greatest user experience, In my opinion, if I had to update pack to be able to get different versions of the API. The same could be said about pack providing 1.0, but maybe I'm not ready for that yet and want to keep using 0.9. If pack removes that from the list, now I have to downgrade pack to select it or go and update my templates manually after creating a new buildpack. 

It would be my preference if this were a free-form field. Perhaps the description could include the most recent API at the time as a hint? Or at least have an "other" option that is free-form. Or is this something the template controls?

Reply:
Aren't the values provided to the end-user derived from the template chosen and therefore managed in a [separate repository](https://github.com/AidanDelaney/cnb-buildpack-templates/blob/main/Go/prompts.toml#L10)? If so, I'm having trouble understanding why `pack` would need to even know about these API versions.